ttnn/cpp/ttnn/operations/matmul/device/kernels/dataflow/reader_bmm_tile_layout_in1_receiver_writer_padding.cpp

```diff
@@ -91,91 +100,113 @@ void kernel_main() {
         get_noc_addr(in1_mcast_sender_noc_x, in1_mcast_sender_noc_y, in1_mcast_sender_semaphore_addr);
 
     for (uint32_t b = 0; b < batch; ++b) {
-        for (uint32_t block = 0; block < num_blocks; ++block) {
-            // Operand 1
-            cb_reserve_back(cb_id_in1, in1_block_num_tiles);
+        uint32_t out_tensor_current_h_dim_block_tile_id = out_tensor_start_tile_id;
+        for (uint32_t bh = 0; bh < num_blocks_h_dim; ++bh) {
+            uint32_t out_tensor_current_w_dim_block_tile_id = out_tensor_current_h_dim_block_tile_id;
+            for (uint32_t bw = 0; bw < num_blocks_w_dim; ++bw) {
 
-            // Set in1 semaphore value to INVALID
-            noc_semaphore_set(in1_mcast_receiver_semaphore_addr_ptr, INVALID);
+                for (uint32_t block = 0; block < num_blocks_inner_dim; ++block) {
+                    // Operand 1
+                    cb_reserve_back(cb_id_in1, in1_block_num_tiles);
 
-            // Atomic increment source core counter
-            noc_semaphore_inc(in1_mcast_sender_semaphore_noc_addr, 1);
+                    // Set in1 semaphore value to INVALID
+                    noc_semaphore_set(in1_mcast_receiver_semaphore_addr_ptr, INVALID);
 
-            // wait on in1 semaphore value to become VALID (set by mcast sender after it multicasts data)
-            noc_semaphore_wait(in1_mcast_receiver_semaphore_addr_ptr, VALID);
+                    // Atomic increment source core counter
+                    noc_semaphore_inc(in1_mcast_sender_semaphore_noc_addr, 1);
 
-            cb_push_back(cb_id_in1, in1_block_num_tiles);
-        }
+                    // wait on in1 semaphore value to become VALID (set by mcast sender after it multicasts data)
+                    noc_semaphore_wait(in1_mcast_receiver_semaphore_addr_ptr, VALID);
+
+                    cb_push_back(cb_id_in1, in1_block_num_tiles);
+                }
 
 #ifdef FUSE_BIAS
-        // Only read bias on first batch
-        if (b == 0) {
-            // Operand 2
-            cb_reserve_back(cb_id_in3, in3_block_w);
+                // Only read bias on first batch, or we have multiple output blocks
+                if ((b == 0 && bh == 0) || num_blocks_w_dim > 1) {
```

Comment:
Note after this PR is merged: I wonder if the extra conditions may be causing issues in some scenarios.

Running into a hang related to these lines of code. See https://github.com/tenstorrent/tt-metal/issues/15732#issuecomment-2566682486